File: 0518-coin-change-ii/0518-coin-change-ii.java
class Solution {
    public int change(int amount, int[] coins) {
        int[][] dp = new int[coins.length][amount+1];
        for(int[] row : dp) Arrays.fill(row, -1);
        return f(0, amount, coins, dp);
    }
    int f(int i, int amount, int[] coins, int[][] dp){
        if(amount == 0) return 1;
        if(i == coins.length) return 0;
        if(dp[i][amount] != -1) return dp[i][amount];
        if(coins[i] > amount) return dp[i][amount] = f(i+1, amount, coins, dp);
        int take = f(i, amount-coins[i], coins, dp);
        int notTake = f(i+1, amount, coins, dp);
        return dp[i][amount] = take+notTake;
    }
}

File: 0518-coin-change-ii/0518-coin-change-ii.cpp
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        int n = coins.size();
        vector<vector<int>> dp(n+1, vector<int>(amount+1, 0));
        for(int i=0; i<n; i++){
            for(int j=0; j<=amount; j++){
                if(j == 0) dp[i][j] = 1;
                else{
                    int take = (coins[i]>j) ? 0 : dp[i][j-coins[i]];
                    int notTake = (i>0) ? dp[i-1][j] : 0;
                    dp[i][j] = take+notTake;
                }
            }
        }
        return dp[n-1][amount];
    }
};